teps/0041-tekton-component-versioning.md

```diff
@@ -0,0 +1,378 @@
+
+### Creating a ConfigMap
+
+Tekton components should create their own ConfigMap which contains
+the version information.
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: pipeline-version
+  namespace: tekton-pipelines
+data:
+  version: v0.19.0
+```
+
+`Role` and `RoleBinding`
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: info
+  namespace: tekton-pipelines
+rules:
+  - apiGroups: [""]
+    resources: ["configmaps"]
+    resourceNames: ["pipelines-version"]
+    verbs: ["get", "describe"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: info
+  namespace: tekton-pipelines
+subjects:
+  - kind: Group
+    name: system:authenticated
+    apiGroup: rbac.authorization.k8s.io
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: info
+```
+
+#### Advantages
+
+1. Other users not having access to the namespace can still have access
+   to the particular ConfigMap.
+
+   `kubectl get cm pipelines-version -n tekton-pipelines`
+
+2. Don't need to create and manage any new types.
+
+#### Disadvantages
+
+1. User need to know the namespace in which Tekton components are installed
+   since the scope is only the namespace.
+
+2. No guarantee that `ConfigMap` and actual installed version will be in sync.
+
+## Test Plan
+
+Basic test scenarios of getting the version information should be added in all the components.
+Other components like CLI can add the test cases based on their usage.
+
+## Design Evaluation
+
+## Drawbacks
+
+## Alternatives
+
+### 1. Provide an Endpoint from the Controller that can be queried
+
+An Endpoint can be created in `controller/main.go` which will return
+the version of Tekton Pipelines and this endpoint can be queried by
+external tools.
+
+#### Advantages
+
+1. It would guarantee that the version matches the installed version.
+2. It will avoid creating another resource such as CRD or ConfigMap
+   and keeping them in sync with the controller.
+
+#### Disadvantages
+
+1. Need to expose that endpoint via `Ingress` or `Route (in OpenShift)`
+   for which user needs to setup the respective controller in their cluster.
+
+   - If we expose the `Service` via `NodePort` then tools like CLI need to be
+     aware of that exposed Service.
+   - If we expose the `Service` via `ClusterIP` then tools like CLI will have to
+     create/maintain a `Pod` which will query that endpoint via the `Service`.
+
+### 2. SSH into the Controller Pod
+
+SSHing inside the controller Pod and then fetch the version from there.
+
+#### Advantages
+
+1. Will not have to maintain extra resource to fetch the version.
+
+#### Disadvantages
+
+1. We should be knowing the namespace in which tekton pipelines is installed.
+2. Appropriate `Role` and `RoleBindings` are required to SSH into the pod.
+
+### 3. Creating a CRD
+
+Tekton components should create a CRD which is to provide information like
+metadata about components.
```

Comment:
Instead of creating, new CRDs, we can consider using already existing CRDs of operator for this purpose.
TektonPipeline, TektonTrigger, TektonDasboard ... which handle the lifecycle of the respective components.

In additon, Operator have also TektonConfig CRD. This CRD is an umbrella CRD which can install 1 or more TektonComponents.

We might be able to achieve what we want by giving access to the status subresource of 'instance' of these CRDs.

Reply:
We cannot totally rely on operator's CRDs as it depends on users how are they installing Pipelines or Triggers.